src/adapter/src/subscribe.rs

```diff
@@ -94,8 +105,105 @@ impl ActiveSubscribe {
+                                                    &right_datums,
+                                                    || Ordering::Equal,
+                                                ) == Ordering::Equal
+                                        }));
+                                    let mut saw_new_row = false;
+                                    for (t, r, d) in group {
+                                        // We don't have to consider (k, v, -1), (k, v, +1) because
+                                        // the rows are already consolidated. For a key, if all we
+                                        // see is retractions we generate (k, null), otherwise we
+                                        // produce all values that k can take. If there are multiple
+                                        // live values for a given key it's impossible for a client
+                                        // to figure out when they are all deleted with ENVELOPE UPSERT.
+                                        if *d < 0 {
+                                            continue;
+                                        }
+                                        let mut packer = row_buf.packer();
+                                        let datums = datum_vec.borrow_with(r);
+                                        for idx in key_indices {
+                                            packer.push(datums[*idx]);
+                                        }
+                                        for idx in 0..self.arity {
+                                            if !key_indices.contains(&idx) {
+                                                packer.push(datums[idx]);
+                                            }
+                                        }
+
+                                        saw_new_row = true;
+                                        new_rows.push((*t, row_buf.clone(), 1));
+                                    }
+
+                                    if !saw_new_row {
+                                        // emit deletion
+                                        let mut packer = row_buf.packer();
+                                        let datums = datum_vec.borrow_with(&start.1);
+                                        for idx in key_indices {
+                                            packer.push(datums[*idx]);
+                                        }
+                                        for _ in 0..self.arity - key_indices.len() {
+                                            packer.push(Datum::Null);
+                                        }
+                                        new_rows.push((start.0, row_buf.clone(), 1));
+                                    }
+                                }
+                                rows = new_rows;
+                            }
+                            SubscribeOutput::Diffs => rows.sort_by_key(|(time, _, _)| *time),
```

Comment:
Might be worth figuring out if we actually need to do this. Pretty sure we don't, but might depend on a guarantee from compute.

Reply:
Yeah, I don't think we do, splitting that question out to https://github.com/MaterializeInc/materialize/pull/18936 in order to get feedback from compute folks